File: src/libs/m3/vfs/VFS.cc
/*
 * Copyright (C) 2015, Nils Asmussen <nils@os.inf.tu-dresden.de>
 * Economic rights: Technische Universitaet Dresden (Germany)
 *
 * This file is part of M3 (Microkernel-based SysteM for Heterogeneous Manycores).
 *
 * M3 is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as
 * published by the Free Software Foundation.
 *
 * M3 is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License version 2 for more details.
 */

#include <base/com/Marshalling.h>
#include <base/stream/Serial.h>

#include <m3/vfs/File.h>
#include <m3/vfs/FileTable.h>
#include <m3/vfs/MountSpace.h>
#include <m3/vfs/VFS.h>
#include <m3/VPE.h>

namespace m3 {

// clean them up after the standard streams have been destructed
VFS::Cleanup VFS::_cleanup INIT_PRIORITY(103);

VFS::Cleanup::~Cleanup() {
    for(fd_t i = 0; i < FileTable::MAX_FDS; ++i)
        delete VPE::self().fds()->free(i);
}

MountSpace *VFS::ms() {
    return VPE::self().mountspace();
}

Errors::Code VFS::mount(const char *path, FileSystem *fs) {
    return ms()->add(new MountSpace::MountPoint(path, fs));
}

void VFS::unmount(const char *path) {
    MountSpace::MountPoint *mp = ms()->remove(path);
    if(mp)
        delete mp;
}

fd_t VFS::open(const char *path, int perms) {
    size_t pos;
    Reference<FileSystem> fs = ms()->resolve(path, &pos);
    if(!fs.valid()) {
        Errors::last = Errors::NO_SUCH_FILE;
        return FileTable::INVALID;
    }
    File *file = fs->open(path + pos, perms);
    if(file) {
        fd_t fd = VPE::self().fds()->alloc(file);
        if(fd == FileTable::INVALID) {
            delete file;
            Errors::last = Errors::NO_SPACE;
        }
        return fd;
    }
    return FileTable::INVALID;
}

void VFS::close(fd_t fd) {
    File *file = VPE::self().fds()->free(fd);
    delete file;
}

Errors::Code VFS::stat(const char *path, FileInfo &info) {
    size_t pos;
    Reference<FileSystem> fs = ms()->resolve(path, &pos);
    if(!fs.valid())
        return Errors::last = Errors::NO_SUCH_FILE;
    return fs->stat(path + pos, info);
}

Errors::Code VFS::mkdir(const char *path, mode_t mode) {
    size_t pos;
    Reference<FileSystem> fs = ms()->resolve(path, &pos);
    if(!fs.valid())
        return Errors::last = Errors::NO_SUCH_FILE;
    return fs->mkdir(path + pos, mode);
}

Errors::Code VFS::rmdir(const char *path) {
    size_t pos;
    Reference<FileSystem> fs = ms()->resolve(path, &pos);
    if(!fs.valid())
        return Errors::last = Errors::NO_SUCH_FILE;
    return fs->rmdir(path + pos);
}

Errors::Code VFS::link(const char *oldpath, const char *newpath) {
    size_t pos1, pos2;
    Reference<FileSystem> fs1 = ms()->resolve(oldpath, &pos1);
    if(!fs1.valid())
        return Errors::last = Errors::NO_SUCH_FILE;
    Reference<FileSystem> fs2 = ms()->resolve(newpath, &pos2);
    if(!fs2.valid())
        return Errors::last = Errors::NO_SUCH_FILE;
    if(fs1.get() != fs2.get())
        return Errors::last = Errors::XFS_LINK;
    return fs1->link(oldpath + pos1, newpath + pos2);
}

Errors::Code VFS::unlink(const char *path) {
    size_t pos;
    Reference<FileSystem> fs = ms()->resolve(path, &pos);
    if(!fs.valid())
        return Errors::last = Errors::NO_SUCH_FILE;
    return fs->unlink(path + pos);
}

void VFS::print(OStream &os) {
    VPE::self().mountspace()->print(os);
}

}

File: src/libs/rustm3/src/vfs/vfs.rs
/*
 * Copyright (C) 2018, Nils Asmussen <nils@os.inf.tu-dresden.de>
 * Economic rights: Technische Universitaet Dresden (Germany)
 *
 * This file is part of M3 (Microkernel-based SysteM for Heterogeneous Manycores).
 *
 * M3 is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as
 * published by the Free Software Foundation.
 *
 * M3 is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License version 2 for more details.
 */

use errors::{Code, Error};
use rc::Rc;
use session::M3FS;
use vfs::{FileInfo, FileMode, FileRef, FSHandle, OpenFlags};
use vpe::VPE;

pub fn mount(path: &str, fs: &str) -> Result<(), Error> {
    let fsobj = match fs {
        "m3fs" => M3FS::new(fs)?,
        _      => return Err(Error::new(Code::InvArgs)),
    };
    VPE::cur().mounts().add(path, fsobj)
}

pub fn unmount(path: &str) -> Result<(), Error> {
    VPE::cur().mounts().remove(path)
}

fn with_path<F, R>(path: &str, func: F) -> Result<R, Error>
                   where F : Fn(&FSHandle, usize) -> Result<R, Error> {
    let (fs, pos) = VPE::cur().mounts().resolve(path)?;
    func(&fs, pos)
}

pub fn open(path: &str, flags: OpenFlags) -> Result<FileRef, Error> {
    with_path(path, |fs, pos| {
        let file = fs.borrow_mut().open(&path[pos..], flags)?;
        VPE::cur().files().add(file)
    })
}

pub fn stat(path: &str) -> Result<FileInfo, Error> {
    with_path(path, |fs, pos| {
        fs.borrow().stat(&path[pos..])
    })
}

pub fn mkdir(path: &str, mode: FileMode) -> Result<(), Error> {
    with_path(path, |fs, pos| {
        fs.borrow().mkdir(&path[pos..], mode)
    })
}

pub fn rmdir(path: &str) -> Result<(), Error> {
    with_path(path, |fs, pos| {
        fs.borrow().rmdir(&path[pos..])
    })
}

pub fn link(old: &str, new: &str) -> Result<(), Error> {
    let (fs1, pos1) = VPE::cur().mounts().resolve(old)?;
    let (fs2, pos2) = VPE::cur().mounts().resolve(new)?;
    if !Rc::ptr_eq(&fs1, &fs2) {
        return Err(Error::new(Code::XfsLink))
    }
    let res = fs1.borrow().link(&old[pos1..], &new[pos2..]);
    res
}

pub fn unlink(path: &str) -> Result<(), Error> {
    with_path(path, |fs, pos| {
        fs.borrow().unlink(&path[pos..])
    })
}
